### PR TITLE
SONARJAVA-4835: Fix FNs on S3242 when using generic types

### PR DESCRIPTION
**Pull request by ADarko22**
Please ensure your pull request adheres to the following guidelines: 

- [ ] Use the following formatting style: [SonarSource/sonar-developer-toolset](https://github.com/SonarSource/sonar-developer-toolset#code-style)
- [ ] Unit tests are passing and you provided a unit test for your fix
- [ ] ITs should pass : To run ITs locally, checkout the [README](https://github.com/SonarSource/sonar-java/blob/master/README.md) of the project.
- [ ] If there is a [Jira](http://jira.sonarsource.com/browse/SONARJAVA) ticket available, please make your commits and pull request start with the ticket number (SONARJAVA-XXXX)

### CODE REVIEW

**Review by kaufco — APPROVED** [review]
LGTM! Maybe add a more complex test case with generics (which I expected to break the implementation, but it didn't)?

```
  public static class Phi {
    public List<? extends Number> getOtherThanOwnerType() {
      return null;
    }
  }

  public static class Theta extends Phi {
    @Override
    public List<Integer> getOtherThanOwnerType() {
      return null;
    }
  }

  public List<Integer> getNonRelatedA(Theta theta) { // Compliant
    return theta.getOtherThanOwnerType();
  }
  
  public List<? extends Number> getNonRelatedB(Theta theta) { // FN
    return theta.getOtherThanOwnerType();
  }
```

**Comment by ADarko22** [comment]
I'm adding those test cases. Note that 
```
  public List<? extends Number> getNonRelatedB(Theta theta) { // FN
    return theta.getOtherThanOwnerType();
  }
```
is not a FN. It is reported as TP :)

**Review by kaufco — CHANGES_REQUESTED** [review]
Sadly, it turned out that my example above indeed creates an FP, which was just masked by a bug that appears when the same method name is used.

As discussed: I'd suggest to close this PR and not to report on types with type parameters.

File: java-checks-test-sources/default/src/main/java/checks/LeastSpecificTypeCheck.java [hunk 1]
```diff
@@ -263,73 +263,4 @@ public BigDecimal getUnaryOperator(UnaryOperator<BigDecimal> func) { // Complian
   public BigDecimal getFunction(Function<BigDecimal, BigDecimal> func) { // S4276 issue to promote UnaryOperator<BigDecimal>
     return func.apply(BigDecimal.ONE);
   }
-
```
Comment (ADarko22) [inline java-checks-test-sources/default/src/main/java/checks/LeastSpecificTypeCheck.java:266]: extracted and re-organized the test samples into a separate sample file

**Review by kaufco — APPROVED** [review]
LGTM! I didn't find a counterexample for the Generics. However, this seems like a lot of test cases. I'll approve anyway, but before merging, maybe you should think about if you can reduce them and which of them really increases the coverage.

**Pull request merged**